src/main/scala/viper/gobra/frontend/Parser.scala

```diff
@@ -116,32 +117,52 @@ object Parser {
       }
     }
 
-    def isErrorFree(parserResults: Vector[Either[Vector[ParserError], PProgram]]): Either[Vector[ParserError], Vector[PProgram]] = {
-      val (errors, programs) = parserResults.partitionMap(identity)
-      if (errors.isEmpty) Right(programs) else Left(errors.flatten)
+    def isErrorFree(parserResults: Map[Path, Either[Vector[ParserError], PProgram]]): Either[Vector[ParserError], Map[Path, PProgram]] = {
+      val (errors, programs) = parserResults.partitionMap({
+        case (_, Left(value)) => Left(value)
+        case (key, Right(value)) => Right((key, value))
+      })
+      if (errors.isEmpty) Right(programs.toMap) else Left(errors.flatten.toVector)
     }
 
-    def samePackage(programs: Vector[PProgram]): Either[Vector[ParserError], Vector[PProgram]] = {
+    def samePackage(pathProgramMap: Map[Path, PProgram]): Either[Vector[ParserError], Map[Path, PProgram]] = {
+      val programs = pathProgramMap.values
       require(programs.nonEmpty)
       val pkgName = programs.head.packageClause.id.name
       val differingPkgNameMsgs = programs.flatMap(p =>
         error(
           p.packageClause,
           s"Files have differing package clauses, expected $pkgName but got ${p.packageClause.id.name}",
-          p.packageClause.id.name != pkgName))
-      if (differingPkgNameMsgs.isEmpty) Right(programs) else Left(pom.translate(differingPkgNameMsgs, ParserError))
+          p.packageClause.id.name != pkgName)).toVector
+      val expectedDir = pathProgramMap.head._1.getParent
+      val differentDirectoriesMsgs = pathProgramMap.flatMap({ case (path, p) =>
+        error(
+          p.packageClause,
+          s"Files are in different directories, expected $expectedDir but got ${path.getParent}",
+         path.getParent != expectedDir)
+      }).toVector
+
+      if (differingPkgNameMsgs.isEmpty && differentDirectoriesMsgs.isEmpty) {
+        Right(pathProgramMap)
+      } else {
+        Left(pom.translate(differingPkgNameMsgs ++ differentDirectoriesMsgs, ParserError))
+      }
     }
 
-    def makePackage(programs: Vector[PProgram]): Either[Vector[ParserError], PPackage] = {
-      val clause = parsers.rewriter.deepclone(programs.head.packageClause)
-      val parsedPackage = PPackage(clause, programs, pom)
+    def makePackage(programs: Map[Path, PProgram]): Either[Vector[ParserError], PPackage] = {
+      val clause = parsers.rewriter.deepclone(programs.head._2.packageClause)
+      val packageDir = programs.head._1.getParent
+
+      val parsedPackage = PPackage(clause, programs.values.toVector, pom, packageDir)
+
       // The package parse tree node gets the position of the package clause:
       pom.positions.dupPos(clause, parsedPackage)
       Right(parsedPackage)
     }
 
     val parsingFn = if (config.cacheParser) { parseSourceCached _ } else { parseSource _ }
-    val parserResults = sources.map(parsingFn)
+    val parserResults = sources.map(src => (Path.of(src.name), parsingFn(src))).toMap
```

Comment:
I don't like the idea of storing a `Source`, inside of a node since this would mean that the contents of the source are stored in memory until that Node is not needed anymore

Reply:
I see. What about having a function that takes a Source and returns a unique identifier (e.g. `SourceIdentifier` that could be a string for now and is the path for file & fromfile sources)?